flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultPreferredLocationsRetrieverTest.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler;
+
+import org.apache.flink.runtime.executiongraph.ExecutionVertex;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+import org.apache.flink.runtime.taskmanager.LocalTaskManagerLocation;
+import org.apache.flink.runtime.taskmanager.TaskManagerLocation;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.hasSize;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests {@link DefaultPreferredLocationsRetriever}.
+ */
+public class DefaultPreferredLocationsRetrieverTest extends TestLogger {
+
+	@Test
+	public void testStateLocationsWillBeReturnedIfExist() {
+		final TaskManagerLocation stateLocation = new LocalTaskManagerLocation();
+
+		final TestingInputsLocationsRetriever.Builder locationRetrieverBuilder =
+			new TestingInputsLocationsRetriever.Builder();
+
+		final ExecutionVertexID consumerId = new ExecutionVertexID(new JobVertexID(), 0);
+		final ExecutionVertexID producerId = new ExecutionVertexID(new JobVertexID(), 0);
+		locationRetrieverBuilder.connectConsumerToProducer(consumerId, producerId);
+
+		final TestingInputsLocationsRetriever inputsLocationsRetriever = locationRetrieverBuilder.build();
+
+		inputsLocationsRetriever.markScheduled(producerId);
+
+		final DefaultPreferredLocationsRetriever locationsRetriever = new DefaultPreferredLocationsRetriever(
+			id -> Optional.of(stateLocation),
+			inputsLocationsRetriever);
+
+		final CompletableFuture<Collection<TaskManagerLocation>> preferredLocations =
+			locationsRetriever.getPreferredLocations(consumerId, Collections.emptySet());
+
+		assertThat(preferredLocations.getNow(null), contains(stateLocation));
```

Comment:
Should we also assert that it contains only `stateLocation` even if producer got a location?

Reply:
The `contains` matcher means it only contains the given collection in the given order.
The name is a bit confusing and I also misunderstood it many times.